fedimint-api-client/src/api/mod.rs

```diff
@@ -912,123 +962,279 @@ impl WsFederationApi<WsClient> {
+        );
+
+        ClientConnection { sender }
+    }
+
+    async fn connection(&self) -> Option<DynClientConnection> {
+        let (sender, receiver) = oneshot::channel();
+
+        self.sender.send(sender).await.ok()?;
+
+        receiver.await.ok()
+    }
+}
+
+#[cfg(all(feature = "enable_iroh", not(target_family = "wasm")))]
+mod iroh {
+    use std::collections::{BTreeMap, BTreeSet};
+
+    use anyhow::anyhow;
+    use async_trait::async_trait;
+    use bitcoin::key::rand::rngs::OsRng;
+    use fedimint_core::module::{ApiError, ApiRequestErased};
+    use fedimint_core::PeerId;
+    use iroh::endpoint::Connection;
+    use iroh::{Endpoint, NodeId, SecretKey};
+    use serde::{Deserialize, Serialize};
+    use serde_json::Value;
+
+    use super::{ApiMethod, DynClientConnection, IClientConnection, IClientConnector};
+
+    const FEDIMINT_ALPN: &[u8] = "FEDIMINT_ALPN".as_bytes();
+
+    #[derive(Debug, Clone)]
+    pub struct IrohConnector {
+        node_ids: BTreeMap<PeerId, NodeId>,
+        endpoint: Endpoint,
+    }
+
+    impl IrohConnector {
+        #[allow(unused)]
+        pub async fn new(peers: BTreeMap<PeerId, NodeId>) -> anyhow::Result<Self> {
+            Ok(Self {
+                node_ids: peers,
+                endpoint: Endpoint::builder()
+                    .discovery_n0()
+                    .secret_key(SecretKey::generate(&mut OsRng))
+                    .alpns(vec![FEDIMINT_ALPN.to_vec()])
+                    .bind()
+                    .await?,
+            })
         }
+    }
 
-        unreachable!();
+    #[async_trait]
+    impl IClientConnector for IrohConnector {
+        fn peers(&self) -> BTreeSet<PeerId> {
+            self.node_ids.keys().copied().collect()
+        }
+
+        async fn connect(&self, peer: PeerId) -> anyhow::Result<DynClientConnection> {
+            let node_id = *self
+                .node_ids
+                .get(&peer)
+                .expect("Could not find node id for peer {peer}");
+
+            let connection = self.endpoint.connect(node_id, FEDIMINT_ALPN).await?;
+
+            Ok(connection.into_dyn())
+        }
+    }
+
+    #[derive(Debug, Clone, Serialize, Deserialize)]
+    struct IrohRequest {
+        method: ApiMethod,
+        request: ApiRequestErased,
+    }
+
+    #[async_trait]
+    impl IClientConnection for Connection {
+        async fn request(
+            &self,
+            method: ApiMethod,
+            request: ApiRequestErased,
+        ) -> anyhow::Result<Value> {
+            let json = serde_json::to_vec(&IrohRequest { method, request })?;
+
+            let (mut sink, mut stream) = self.open_bi().await?;
+
+            sink.write_all(&json).await?;
+
+            sink.finish()?;
+
+            let response = stream.read_to_end(1_000_000).await?;
+
+            let response = serde_json::from_slice::<Result<Value, ApiError>>(&response)?;
+
+            response.map_err(|e| anyhow!("Api Error: {:?}", e))
```

Comment:
`.context(...)` with anyhow, much better than stringly concatendated messages.

Reply:
BTW. `{:?}` prints the whole backtrace in the formatting on anyhow errors, and is generally ugly on all types, leading to meh error messages, so should be generally avoided when dealing with errors.